### ISSUE TITLE
Can be the esp8266 server and client at  the same time? 

### ISSUE THREAD

**Issue by amprodes** [Links2004/arduinoWebSockets#488]

**Comment by Links2004** [comment]
yes it can, only local loop back connection is not supported (TCP stack limitation).

see #136 and #401 for more details and example code

**Comment by amprodes** [comment]
Mine is not working...

this is my code... I'm using wifi as AP...

`#include <Arduino.h>
#include <ESP8266WiFi.h>  
#include <WebSocketsServer.h> 
#include <WebSocketsClient.h>
#include <Hash.h>  

#ifndef APSSID
#define APSSID "amElectronics"
#define APPSK  "bin2bin2a"
#endif

WebSocketsServer webSocket(81);  // Create a websocket server on port 1884
WebSocketsClient webSocketCli;

/* Set these to mine desired credentials. */
  const char *ssid      = APSSID;
  const char *password  = APPSK;
  boolean     Master    = true;

/*___________________________________________SETUP_FUNCTIONS__________________________________________________________*/

  void event(const char * payload, size_t length) {
    Serial.printf("got message: %s\n", payload);
  }

  void startWiFi() { // Start a Wi-Fi access point, and try to connect to some given access points. Then wait for either an AP connection
    Serial.println();
    Serial.print("Setting WIFI_AP … ");
    Serial.println(WiFi.mode(WIFI_AP) ? "Ready" : "Failed!");
    Serial.print("Setting soft-AP … ");
    Serial.println(WiFi.softAP(ssid, password) ? "Ready" : "Failed!");
    Serial.print("Soft-AP IP address =");
    Serial.println(WiFi.softAPIP());
    
}

  void startWebSocket() { // Starts a WebSocket server
    webSocket.begin();                          // start the websocket server
    webSocket.onEvent(webSocketEvent);          // if there's an incomming websocket message, go to function 'webSocketEvent'

  }
  
  void startWebSocketClient() { // Starts a WebSocket server
    webSocketCli.begin("192.168.4.1", 81, "/");
    webSocketCli.onEvent(webSocketEventCli);      // if there's an incomming websocket message, go to function 'webSocketEvent'
    // try ever 5000 again if connection has failed
    webSocketCli.setReconnectInterval(5000);
    
    // start heartbeat (optional)
    // ping server every 15000 ms
    // expect pong from server within 3000 ms
    // consider connection disconnected if pong is not received 2 times
    webSocketCli.enableHeartbeat(15000, 3000, 2);
  }
/*__________________________________________________________SETUP_FUNCTIONS__________________________________________________________*/

/*__________________________________________________________HELPER_FUNCTIONS__________________________________________________________*/

void webSocketEvent(uint8_t num, WStype_t type, uint8_t * payload, size_t lenght) { // When a WebSocket message is received

   switch (type) {
    case WStype_DISCONNECTED:             // if the websocket is disconnected
      Serial.printf("[%u] Disconnected!\n", num);
      break;
    case WStype_CONNECTED: {              // if a new websocket connection is established
        IPAddress ip = webSocket.remoteIP(num);
        Serial.printf("\"[%u] Connected from %d.%d.%d.%d url: %s\n \"", num, ip[0], ip[1], ip[2], ip[3], payload);
        // send message to client
        webSocket.sendTXT(num, " Conectado");
      }
      break;
    case WStype_TEXT:                     // if new text data is received
      Serial.printf("[%u] get Text: %s\n", num, payload);
      webSocket.sendTXT(num, payload);
      break;
    // For everything else: do nothing
    case WStype_ERROR:
      Serial.printf("[%u] get Error: %s\n", num, payload);
      break;
    case WStype_FRAGMENT_TEXT_START:
    case WStype_FRAGMENT_BIN_START:
    case WStype_FRAGMENT:
    case WStype_FRAGMENT_FIN:
    case WStype_BIN: 
    default:
      break;
  }
}

void webSocketEventCli(WStype_t type, uint8_t * payload, size_t length) {

  switch(type) {
    case WStype_DISCONNECTED:
      Serial.printf("[WSc] Disconnected!\n");
      break;
    case WStype_CONNECTED: 
    {
      Serial.printf("[WSc] Connected to url: %s\n", payload);

      // send message to server when Connected
      webSocketCli.sendTXT("Clien from Esp8266 is Connected");
    }
      break;
    case WStype_TEXT:
      Serial.printf("[WSc] get text: %s\n", payload);

      // send message to server
       webSocketCli.sendTXT("got message too");
      break;
    case WStype_BIN:
      Serial.printf("[WSc] get binary length: %u\n", length);
      hexdump(payload, length);

      // send data to server
      // webSocket.sendBIN(payload, length);
      break;
        case WStype_PING:
            // pong will be send automatically
            Serial.printf("[WSc] get ping\n");
            break;
        case WStype_PONG:
            // answer to a ping we send
            Serial.printf("[WSc] get pong\n");
            break;
    }

}
/*__________________________________________________________HELPER_FUNCTIONS__________________________________________________________*/

void setup() {
  
  Serial.begin(115200);  
  delay(5000);

    Serial.println();
 
    startWiFi();                 // Start a Wi-Fi access point, and try to connect to some given access points. Then wait for either an AP or STA connection
    startWebSocket();            // Start a WebSocket server
    startWebSocketClient();
    
    for(uint8_t t = 4; t > 0; t--) {
        Serial.printf("[SETUP] BOOT WAIT %d...\n", t);
        Serial.flush();
        delay(1000);
    }
 }

unsigned long messageTimestamp = 0; 

void loop() { 

  webSocket.loop(); 
  //webSocketCli.loop();
   
  delay(50); // Give the ESP time to handle network...
  
  webSocketCli.sendTXT("..."); 
  if (Serial.available()) {
    Serial.setTimeout(100);
    String s = Serial.readStringUntil('\n');
    //Serial.write(s.c_str());
    Serial.printf(s.c_str());
    if (s.startsWith("node: ")) {
      String data;
      int end = s.indexOf('\r');
      if (end > 0) {
        data = s.substring(6, end);
      } else {
        data = s.substring(6);
      }
      
      webSocketCli.sendTXT(data.c_str());
     
    }
  }
}`

please help!

**Comment by sovcik** [comment]
Could you attach a log?
